src/services/identity/CollaboratorsService.ts

```diff
@@ -0,0 +1,240 @@
+    this.whitelist = whitelist
+  }
+
+  deriveAllowedRoleFromEmail = async (fullEmail: string) => {
+    const whitelistEntries = await this.whitelist.findAll({
+      where: {
+        expiry: {
+          [Op.or]: [{ [Op.is]: null }, { [Op.gt]: new Date() }],
+        },
+      },
+    })
+
+    const matchedDomains = whitelistEntries.filter((entry) =>
+      fullEmail.endsWith(entry.email)
+    )
+
+    if (!matchedDomains.length) return null
+
+    // TODO: Modify this method because the presence of the expiry field is not
+    // the best way of differentiating Admin/Contributor roles
+    return matchedDomains[0].expiry
+      ? CollaboratorRoles.Contributor
+      : CollaboratorRoles.Admin
+  }
+
+  list = async (siteName: string, requesterId?: string) => {
+    // Note:
+    // ===============================================
+    // We need to query from the Sites table instead of the SiteMembers table
+    // because Sequelize only recognizes that there is a relationship between Sites <-> Users.
+    // This means that we cannot retrieve joins if we start the query in the SiteMembers table.
+    //
+    // However, the converse is possible, i.e. we can query the Sites table and retrieve joined
+    // records from the Users table, along with the SiteMember records.
+    const site = await this.siteRepository.findOne({
+      where: { name: siteName },
+      include: [
+        {
+          model: User,
+          as: "site_members",
+        },
+      ],
+    })
+    const collaborators = site?.site_members ?? []
+
+    // We perform the following sort via application code because:
+    // - sorting it via the ORM code alone is quite complicated
+    // - putting the sorting logic into a stored SQL function involves DB migration work
+    // - we can achieve this easily with lodash, and there is unlikely to be a performance hit
+    //   given the small number of collaborators in each site
+    return _.orderBy(
+      collaborators,
+      [
+        // Prioritize Admins over Contributors
+        (
+          collaborator: User & {
+            SiteMember: SiteMember
+          }
+        ) => collaborator.SiteMember.role === CollaboratorRoles.Admin,
+        // Prioritize elements where the userId matches the requesterId (i.e. "you")
+        (
+          collaborator: User & {
+            SiteMember: SiteMember
+          }
+        ) => collaborator.id.toString() === requesterId,
+        // Prioritize the last logged in user
+        (
+          collaborator: User & {
+            SiteMember: SiteMember
+          }
+        ) => collaborator.lastLoggedIn,
+      ],
+      ["desc", "desc", "asc"]
+    )
+  }
+
+  create = async (siteName: string, email: string, acknowledged: boolean) => {
+    let site
+    let user
+
+    if (!email || !validator.isEmail(email)) {
+      return new BadRequestError(
+        "That doesn't look like a valid email. Try a gov.sg or other whitelisted email."
+      )
```

Comment:
Should we throw the error directly here instead of letting it be handled at the router level?

Reply:
My thought here was that we should move towards a [neverthrow](https://github.com/supermacro/neverthrow)-like pattern (but I didn't want to introduce neverthrow into this specific PR just yet - we should do it in a separate PR to prevent scope creep). 

In this pattern, we want to ensure that library methods don't throw errors themselves as far as possible. Instead, we want to get them to return errors of specific types so that their callers can figure out what to do with them. 

In this PR, because we're not introducing neverthrow yet:
- the business-logic/expected errors (e.g. site not found) will be returned as a specific error
- the unexpected errors (e.g. database-related errors) will be thrown

In a separate PR, when we introduce neverthrow, errors in both categories will be returned as specific errors.

@seaerchin - thoughts?